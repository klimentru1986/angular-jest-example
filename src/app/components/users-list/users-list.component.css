th,
td {
  padding: 5px 10px;
  text-align: center;
}
