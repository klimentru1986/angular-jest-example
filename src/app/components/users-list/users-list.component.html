<table>
  <thead>
    <tr>
      <th>Идентификатор</th>
      <th>Имя</th>
      <th>Возраст</th>
    </tr>
  </thead>
  <tbody>
    <tr *ngFor="let user of users$ | async">
      <td>{{user.id}}</td>
      <td>{{user.name}}</td>
      <td>{{user.age}}</td>
    </tr>
  </tbody>
</table>

<hr/>

<div *ngIf="users$ | async as users">
  <app-user-details [userInfo]="users[0]"></app-user-details>
</div>

