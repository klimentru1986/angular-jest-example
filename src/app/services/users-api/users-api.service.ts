import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs/Observable';
import { User } from '../../models/user.model';

@Injectable()
export class UsersApiService {
  constructor(private http: HttpClient) {}

  getUsers(): Observable<User[]> {
    const url = 'https://5a2762ec84a17f0012a945b2.mockapi.io/persons';

    return this.http.get<User[]>(url);
  }
}
