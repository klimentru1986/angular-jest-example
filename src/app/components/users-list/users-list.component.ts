import { Component, OnInit } from '@angular/core';
import { Observable } from 'rxjs';
import { User } from '../../models/user.model';
import { UsersApiService } from '../../services/users-api/users-api.service';

@Component({
  selector: 'app-users-list',
  templateUrl: './users-list.component.html',
  styleUrls: ['./users-list.component.css']
})
export class UsersListComponent implements OnInit {
  public users$: Observable<User[]>;

  constructor(private usersApi: UsersApiService) {}

  ngOnInit() {
    this.users$ = this.usersApi.getUsers();
  }
}
