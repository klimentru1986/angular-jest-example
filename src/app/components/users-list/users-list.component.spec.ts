import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { UsersApiService } from '../../services/users-api/users-api.service';
import { UsersListComponent } from './users-list.component';

import { of } from 'rxjs/observable/of';
import { User } from '../../models/user.model';

describe('UsersListComponent', () => {
  let comp: UsersListComponent;
  let fixture: ComponentFixture<UsersListComponent>;
  let expectedUsers: User[];

  beforeEach(() => {
    expectedUsers = [
      { id: '5a276e9b174e685a77674d49', age: 38, name: 'Pruitt Ramirez' },
      { id: '5a276e9b6561a2e06f682b9c', age: 33, name: 'Tabitha Dickson' }
    ];

    const usersApiServiceStub = {
      getUsers: () => {}
    };
    TestBed.configureTestingModule({
      declarations: [UsersListComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [{ provide: UsersApiService, useValue: usersApiServiceStub }]
    });
    fixture = TestBed.createComponent(UsersListComponent);
    comp = fixture.componentInstance;
  });

  it('can load instance', () => {
    expect(comp).toBeTruthy();
  });

  it('makes expected calls', () => {
    const usersApiServiceStub: UsersApiService = fixture.debugElement.injector.get(UsersApiService);
    spyOn(usersApiServiceStub, 'getUsers');
    fixture.detectChanges();

    expect(usersApiServiceStub.getUsers).toHaveBeenCalled();
  });

  it('should render mock users', async () => {
    const el: HTMLElement = fixture.debugElement.nativeElement.querySelector('table');
    const usersApiServiceStub: UsersApiService = fixture.debugElement.injector.get(UsersApiService);
    jest.spyOn(usersApiServiceStub, 'getUsers').mockReturnValue(of(expectedUsers));

    fixture.detectChanges();
    const elHtml = el.innerHTML;

    expect(elHtml).toContain(expectedUsers[0].id);
    expect(elHtml).toContain(expectedUsers[0].name);
    expect(elHtml).toContain(expectedUsers[0].age);
    expect(elHtml).toContain(expectedUsers[1].id);
    expect(elHtml).toContain(expectedUsers[1].name);
    expect(elHtml).toContain(expectedUsers[1].age);
  });
});
