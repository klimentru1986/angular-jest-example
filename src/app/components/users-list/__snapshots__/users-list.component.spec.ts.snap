// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersListComponent should renders correctly 1`] = `
<div
  id="root3"
  ng-version="8.0.0"
>
  <table>
    <thead>
      <tr>
        <th>
          Идентификатор
        </th>
        <th>
          Имя
        </th>
        <th>
          Возраст
        </th>
      </tr>
    </thead>
    <tbody>
      
      <tr>
        <td>
          5a276e9b174e685a77674d49
        </td>
        <td>
          Pruitt Ramirez
        </td>
        <td>
          38
        </td>
      </tr>
      <tr>
        <td>
          5a276e9b6561a2e06f682b9c
        </td>
        <td>
          Tabitha Dickson
        </td>
        <td>
          33
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  
  <div>
    <app-user-details
      ng-reflect-user-info="[object Object]"
    />
  </div>
</div>
`;
